### ISSUE TITLE
Register application/jsonld alongside application/ld+json?

### ISSUE THREAD

**Issue by iherman** [w3c/json-ld-syntax#287]
This issue came up in the DID Working Group, as part of the discussion on how to define a media type for a DID document. Instead of repeating all the issues, please refer to https://github.com/w3c/did-spec/issues/1.

If accepted it should be done before CR, because that should be reflected in the API document as well.

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-10-25-json-ld#section3-2).

- `ACTION:  find reference to file extensions for profiles re annotation (Rob Sanderson)`
- `ACTION:  comment in issue inviting DID WG to JSON-LD WG regular call next week to discuss (Rob Sanderson)`
<details><summary><i class="differentiate">View the transcript</i></summary>
pplication/jsonld issue #287<br/>
Rob Sanderson: <a href="https://github.com/w3c/json-ld-syntax/issues/287">https://github.com/w3c/json-ld-syntax/issues/287</a><br/>
Ivan Herman: am representing the DID WG<br/>
… it is in that capacity that I raised this question<br/>
Dave Longley: this might be better if people could rad over the DID spec issue first,<br/>
… but it’s not clear how to handle mimetypes<br/>
… this is documented at the discussion there (at DID)<br/>
Pierre-Antoine Champin: the goal is to be able to declare application/did+jsonld?<br/>
… ?<br/>
Ivan Herman: Yes. It’s unclear whether application/did+ld+json is acceptable<br/>
Dave Longley: no one does this^ … but it doesn’t seem illegal … and questionable what tools would do.<br/>
Ivan Herman: there is no current registered mimetype that uses two + signs<br/>
… so we are not sure if this is really possible<br/>
Dave Longley: VC WG did not choose a mimetype<br/>
Dave Longley: VCWG didn’t pick a MIME type that i can see<br/>
Ivan Herman: that is a typical case where this might make sense<br/>
Rob Sanderson: <a href="https://www.w3.org/TR/annotation-protocol/#annotation-retrieval">https://www.w3.org/TR/annotation-protocol/#annotation-retrieval</a><br/>
Rob Sanderson: for Annotations, we went with a profile<br/>
… you can trigger a file extension from a profile, says IANA. If so, that would fulfill the requirements listed on the DID issue comment<br/>
… but I don’t have a reference for that<br/>
… we would have to go search the minutes<br/>
Rob Sanderson: <a href="https://www.iana.org/assignments/profile-uris/profile-uris.xhtml">https://www.iana.org/assignments/profile-uris/profile-uris.xhtml</a><br/>
Ivan Herman: we defined no extension for annotation<br/>
Action #3: find reference to file extensions for profiles re annotation (Rob Sanderson)<br/>
Rob Sanderson: if we did the application/jsonld, processors would have to know about it, which would be a normative change to the document<br/>
Ivan Herman: it would change the documents to allow this new media type in parallel to the older one<br/>
… it is simple for the document, but causes a lot of change elsewhere, because JSON-LD processors would have to change, too.<br/>
Rob Sanderson: we aren’t going to decide this today<br/>
… if there is no change here, we are on-track for CR<br/>
… a change here would push CR back by at least 2 weeks with Thanksgiving looming<br/>
Dave Longley: we want this group to enable other groups to be able to extend jsonld MIME type w/ file extension support – just a question of whether we need to register application/jsonld or not … either way should provide some advice to extenders on what to do<br/>
Dave Longley: we did discuss this in the 1.0 days, but it might not recorded in issues<br/>
Rob Sanderson: <a href="https://github.com/w3c/did-spec/issues/1#issuecomment-545103700">https://github.com/w3c/did-spec/issues/1#issuecomment-545103700</a><br/>
Rob Sanderson: if you can register a file extension to a profile, than that is what we could recommend<br/>
… how does application/jsonld solve the problem?<br/>
Ivan Herman: because then you can do application/did+jsonld<br/>
… no multiple plus signs<br/>
Rob Sanderson: this came up in XML days (application/xhtml+xml)<br/>
Ivan Herman: i couldn’t find any example of this (multiple plus signs) in IANA<br/>
… if I have a doc that is application/did+ld+json, and a JSON-LD proc, then that proc should handle it fine regardless of the DID aspect.<br/>
… if I do application/did+json, then a JSON-LD proc may or may not work correctly<br/>
… (handle it as JSON-LD)<br/>
Pierre-Antoine Champin: at the discussion at the DID WG, there is discussion of an alias<br/>
… is this an official construct of IANA?<br/>
Ivan Herman: I don’t know, but i think not; it just reflects that two media types may share suffix (but I may be wrong)<br/>
… I propose that azaroth digs up the remark from IANA about file extensions and we’d come back to the issue<br/>
Pierre-Antoine Champin: a profile approach could look pretty complicated to some people<br/>
… it might be a shame, in terms of marketing<br/>
Dave Longley: +1<br/>
Rob Sanderson: dlongley, the two WGs can meet at the next JSON-LD WG meeting<br/>
… to discuss this<br/>
Action #4: comment in issue inviting DID WG to JSON-LD WG regular call next week to discuss (Rob Sanderson)<br/>
</details>

**Comment by gkellogg** [comment]
See [my comment](https://github.com/w3c/json-ld-wg/issues/123#issuecomment-546991399) in the WG issue.

> 👎 to registering `application/jsonld`, if there is a way to avoid it. Registering with a profile parameter, or something like `application/anno-ld+json` would seem preferable. What happens if there are sub-profiles of anno, or some other sub-type?
> 
> Surely, we can't be the first format to run-into more than one-level of media-type sub-types.

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-11-01-json-ld#section3).

- `RESOLVED:  Ask IETF if the JSON-LD media type registration can specify that it should be able to be used with additional +s, such as did+ld+json, with the intent to fallback to ld+json and then to json. If the answer is no, then we proceed with the profile pattern.`
- `RESOLVED:  In IANA considerations, allow file extension registration with profile registration.`
- `RESOLVED:  Conforming json-ld processors SHOULD treat *+ld+json in the same way as the application/ld+json media type`
- `ACTION:  contact IETF about the multiple pluses in the JSON-LD registration (Rob Sanderson)`
<details><summary><i class="differentiate">View the transcript</i></summary>
Media Types for JSON-LD<br/>
Rob Sanderson: <a href="https://github.com/w3c/json-ld-syntax/issues/287">https://github.com/w3c/json-ld-syntax/issues/287</a><br/>
Ivan Herman: See related DID issue<br/>
Rob Sanderson: the DID WG is looking for how to have a media-type, with a specific file extension<br/>
… which would be recognized as a subtype of JSON-LD<br/>
… What would happen with multiple + signs, like application/did+ld+json ?<br/>
… Or should it be application/did+jsonld<br/>
… Or could an extension be registered with a profile of application/ld+json?<br/>
Gregg Kellogg: did anyone consider application/did-ld+json? Does it satisfy any of the concern?<br/>
Ivan Herman: that was one of the options.<br/>
Markus Sabadello: we want a specific mime-type, in order to have a file extension,<br/>
… and also to be able to specify the semantics of fragments in our format.<br/>
… Problem seems to be that we are defining a subtype of a subtype.<br/>
Dave Longley: potential options: application/did-ld+json, application/did+ld+json application/ld+json;profile=URI, application/did+jsonld, application/did+json<br/>
Markus Sabadello: The hyphen or the multiple-plus solutions would work I think.<br/>
Rob Sanderson: about the extension, there was a discussion in the Annotation WG;<br/>
… (-> see also a separate action <a href="https://github.com/w3c/json-ld-wg/issues/123">https://github.com/w3c/json-ld-wg/issues/123</a> )<br/>
… we checked with IETF and IANA (Mark Nottigham).<br/>
… We asked if a file extension could be registered with a profile parameter,<br/>
… without a new media-type.<br/>
… The answer was yes, provided that the “media-type owner” acknowledge it.<br/>
… In this case, the “media-type owner” is us (the JSON-LD WG).<br/>
… To me, the profile pattern is the preferable way. We should explore this first.<br/>
Rob Sanderson: It would look like: application/ld+json;profile="<a href="https://w3.org/ns/did"">https://w3.org/ns/did"</a> (or whatever URI)<br/>
Rob Sanderson: Profile parameter: <a href="https://www.w3.org/TR/json-ld11/#iana-considerations">https://www.w3.org/TR/json-ld11/#iana-considerations</a><br/>
Benjamin Young: <a href="https://tools.ietf.org/html/rfc6906">https://tools.ietf.org/html/rfc6906</a> - The ‘profile’ Link Relation Type<br/>
Benjamin Young: <a href="https://tools.ietf.org/html/rfc7284">https://tools.ietf.org/html/rfc7284</a> - The Profile URI Registry RFC<br/>
Rob Sanderson: <a href="https://www.iana.org/assignments/profile-uris/profile-uris.xhtml">https://www.iana.org/assignments/profile-uris/profile-uris.xhtml</a> - Profile URI registry:<br/>
Ted Thibodeau Jr: Can anyone point to a documentation of the profile suffix on mime-types?<br/>
… We have been discussing the “+” solution, because it allows a fallback.<br/>
… But according to the RFC, only the final + matters for this fallback mechanism.<br/>
… But we hope that implementations consider all + signs,<br/>
… otherwise, application/did+ld+json would fallback directly to JSON,<br/>
… which is not ideal.<br/>
… I would suggest that application/jsonld be declared, as an “alias” to application/ld+json,<br/>
… i.e. having the same file extension.<br/>
Ivan Herman: See reference to the ‘+’ descriptions<br/>
Dave Longley: the JSON-LD should advertise the possibility to associate file extensions to profiles,<br/>
… but the tooling is missing for properly working with the profile approach. (e.g., npm mime package)<br/>
Markus Sabadello: about the fragments,<br/>
… the way we are using the fragments in DID should be compatible with the way they are used in JSON-LD,<br/>
… so the fallback mentioned above should work to some extent.<br/>
… A pure JSON-LD tool should be able to process the fragments in a DID document.<br/>
Dave Longley: +1 we absolutely want it to process the fragment the same way<br/>
Ivan Herman: I don’t think the fragment ID is defined for the JSON format, is there one for JSON-LD?<br/>
Gregg Kellogg: I think we say they have the same semantics as in RDF.<br/>
“Fragment identifiers used with application/ld+json are treated as in RDF syntaxes, as per RDF 1.1 Concepts and Abstract Syntax [RDF11-CONCEPTS].”<br/>
Benjamin Young: JSON does not have fragment identifiers.<br/>
Gregg Kellogg: is there inheritability of fragment identifiers across media-types?<br/>
… It would be logical.<br/>
Ted Thibodeau Jr: application/did-ld+json would need to echo much of application/ld+json<br/>
Ted Thibodeau Jr: application/did+ld+json could inherit if MIME RFC made all + important<br/>
Ted Thibodeau Jr: application/did+jsonld would inherit/fallback to application/jsonld<br/>
Rob Sanderson: as TallTed pointed out, there is some time pressure.<br/>
… We are close to recommendation.<br/>
… Would some of the options require a normative change in JSON-LD?<br/>
… Consensus seems to be that we should say, in the IANA section,<br/>
… that profile IRIs can be associated to new file extensions.<br/>
… did[+-]ld+json would not require any change.<br/>
… did+jsonld would require a change.<br/>
… Let’s focus on the options requiring some change.<br/>
Dave Longley: could the application/ld+json registration possibly talk about additional ‘+’s? or would that be too out of scope?<br/>
Ted Thibodeau Jr: application/jsonld parallel to application/ld+json with the latter preferred for basic JSON-LD would work … only derivatives of JSON-LD (like DID) would use the former … but maybe did-ld should be considered as parallel to ld as derivatives of JSON<br/>
Ivan Herman: the registration of did+ld+json only concerns the DID WG, won’t affect the JSON-LD WG.<br/>
… re: did-ld+json, my impression is that this mime-type would be considered independent from ld+json<br/>
… The reason why it came up, if I’m correct, was to define the semantics of fragment identifiers.<br/>
… We just pointed out that JSON-LD does define a semantics for fragment identifiers.<br/>
… Is DID compatible with it? If not, then it should not be a subtype of ld+json,<br/>
… regardless of the possibility for IANA.<br/>
Markus Sabadello: +1, in DID documents we want to use standard JSON-LD fragment behavior<br/>
Dave Longley: for me, they should be compatible, and I think this is the opinion of the DID WG.<br/>
… I wonder if we could not just mention in the IANA declaration, that +ld+json should be considered a subtype of ld+json,<br/>
… even despite what the RFC says about the final +.<br/>
Rob Sanderson: We can ask IANA what they would think about that if we did that.<br/>
Ted Thibodeau Jr: concerned about timing<br/>
Markus Sabadello: I agree with Dave about the standard fragment semantics.<br/>
Rob Sanderson: dlongley, you mentioned problems with the tooling for the profile approach.<br/>
… can tools be updated more quickly than specs?<br/>
Dave Longley: I’m not sure they can.<br/>
… There are many toolchains out there…<br/>
Ted Thibodeau Jr: if the only concern is the fragment semantics,<br/>
… and they appear to be in synchrony,<br/>
… do we need a DID media-type in the 1st place?<br/>
Dave Longley: we also want a file extension.<br/>
Ted Thibodeau Jr: multiple file extensions can be associated with a single media-type.<br/>
… Why not associate the .did extension to application/ld+json?<br/>
Rob Sanderson: this would be a bad precedent<br/>
Ted Thibodeau Jr: there are lot of things out there, sharing the same media-type,<br/>
… with very different contents.<br/>
Benjamin Young: I’m assuming you want a file extension so that desktop apps do the right thing (double click)./<br/>
… Most application do not trust only the extension, and check the content once opened.<br/>
… We could register several extensions with application/ld+json, applications should still check the content.<br/>
… The same problem exists with JSON.<br/>
Markus Sabadello: we also want to be able to do content negotiation<br/>
… we want to be able to distinguish btw a DID document and another JSON-LD document.<br/>
Rob Sanderson: can we update the IANA section, saying that<br/>
… when a file extension is associated with a profile,<br/>
Ivan Herman: this discussion started with the question about multiple +. Shouldn’t we go back to checking this? If we can, this is the simplest solution.<br/>
… Let’s check with Heather (from IETF). If the answer is no, then we can look for another solution.<br/>
Dave Longley: +1 to ivan<br/>
Benjamin Young: +1 to ivan<br/>
Gregg Kellogg: +1 to ivan<br/>
Ted Thibodeau Jr: did+ld+json conforms to RFC. but tools will probably fall back to json, not to ld+json. which may be OK.<br/>
Markus Sabadello: but my understanding is that did+ld+json would NOT automatically inherit from ld+json<br/>
Rob Sanderson: We could require that the registration of a file extension be combined with the registration of a profile URI. Then there’s the express intent of the profile URI and the extension being linked. Then the registration for ld+json would explicitly inherit all of the file extensions of all of the profiles.<br/>
… the RFC says that only the last + is relevant.<br/>
… I agree that we should check. I’m afraid the answer is no.<br/>
Ivan Herman: Let’s ask, and prepare a fallback plan.<br/>
Rob Sanderson: I think the fallback should be profile.<br/>
Benjamin Young: the did+ld+json IANA declaration could explicitly state that it inherits ld+json.<br/>
Proposed resolution: Ask IETF if the JSON-LD media type registration can specify that it should be able to be used with additional +s, such as did+ld+json, with the intent to fallback to ld+json and then to json. If the answer is no, then we proceed with the profile pattern. (Rob Sanderson)<br/>
Gregg Kellogg: +1<br/>
Dave Longley: +1<br/>
Rob Sanderson: +1<br/>
Pierre-Antoine Champin: +1<br/>
Ivan Herman: +1<br/>
Benjamin Young: +1<br/>
David I. Lehn: +1<br/>
Ted Thibodeau Jr: +1<br/>
Ted Thibodeau Jr: worst case, did+ld+json is equivalent to did-ld+json … and things need to be explicitly and/or redundantly stated in the DID space.<br/>
Ted Thibodeau Jr: this should require less retooling than reliance on profile – which currently only works in conneg<br/>
Resolution #2: Ask IETF if the JSON-LD media type registration can specify that it should be able to be used with additional +s, such as did+ld+json, with the intent to fallback to ld+json and then to json. If the answer is no, then we proceed with the profile pattern.<br/>
Markus Sabadello: if the answer is no in general, we can decide later if we go for the profile pattern, or something else…<br/>
Markus Sabadello: +1 to above proposal, but fallback could still be either profile, or did+ld+json<br/>
Action #1: contact IETF about the multiple pluses in the JSON-LD registration (Rob Sanderson)<br/>
Benjamin Young: <a href="https://tools.ietf.org/html/rfc7231#section-5.3.2">https://tools.ietf.org/html/rfc7231#section-5.3.2</a><br/>
Dave Longley: the did+ld+json is legal; even if IETF decides that it does NOT inherit ld+json in general, we can bake this inheritance in this specific media-type<br/>
Benjamin Young: we should check JS libraries, see how they handle this<br/>
Ted Thibodeau Jr: JSON-LD WG could suggest that unrecognized /…+ld+json be treatable as /ld+json<br/>
… (this could be non-normative at this point)<br/>
Dave Longley: +1 to TallTed<br/>
Benjamin Young: +1 to TallTed<br/>
Markus Sabadello: +1 to TallTed<br/>
Rob Sanderson: re. TallTed suggestion, it boils down to what IETF thinks of that pattern<br/>
Ted Thibodeau Jr: even if they disagree about it in general, we can still recommend it for JSON-LD<br/>
Gregg Kellogg: +1 to pchampin<br/>
Markus Sabadello: +1 to pchampin<br/>
Dave Longley: +1<br/>
Ted Thibodeau Jr: We could recommend that JSON-LD processors process application/*+ld+json<br/>
Proposed resolution: In IANA considerations, allow file extension registration with profile registration. (Rob Sanderson)<br/>
Rob Sanderson: +1<br/>
Pierre-Antoine Champin: +1<br/>
Gregg Kellogg: +1<br/>
Dave Longley: +1<br/>
Markus Sabadello: +1<br/>
Ivan Herman: +1<br/>
Resolution #3: In IANA considerations, allow file extension registration with profile registration.<br/>
*Proposed resolution: Conforming json-ld processors SHOULD treat *+ld+json in the same way as the application/ld+json media type** *(Rob Sanderson)<br/>
Dave Longley: +1<br/>
Markus Sabadello: +1<br/>
Gregg Kellogg: +1<br/>
Rob Sanderson: +1<br/>
Pierre-Antoine Champin: +1<br/>
Ivan Herman: +1<br/>
Ted Thibodeau Jr: +1<br/>
Resolution #4: Conforming json-ld processors SHOULD treat *+ld+json in the same way as the application/ld+json media type<br/>
Benjamin Young: +1<br/>
Rob Sanderson: I hope we get a response from IETF for next week.<br/>
Benjamin Young: +1<br/>
Ivan Herman: anyway, we can put in the document what we have resolved, regardless.<br/>
Dave Longley: +1 this works regardless<br/>
Markus Sabadello: thank you all for taking the time today!<br/>
</details>

**Comment by iherman** [comment]
I agree to close this.
